### PR TITLE
#893187 - support multiple dSYMs upload

### PR DESCRIPTION
**Pull request by AlexRukhlin**

### CODE REVIEW

File: Tasks/VSMobileCenterUpload/vsmobilecenterupload.ts [hunk 1]
```diff
@@ -331,7 +330,32 @@ async function run() {
...
+            {
+                if (dsymFolder) {
+                    let path = utils.checkAndFixFilePath(dsymFolder, "symbolsPath", continueIfSymbolsNotFound);
+                    if (path) {
+                        symbolsPaths.push(path);
+                    }
+                }
+            })
+        } else {
+            let simbolsFile = utils.resolveSinglePath(symbolsPathPattern, continueIfSymbolsNotFound, packParentFolder);
```
Comment (DavidStaheli) [inline Tasks/VSMobileCenterUpload/vsmobilecenterupload.ts]: "simbolsFile" -> "symbolsFile"

Comment (AlexRukhlin) [inline Tasks/VSMobileCenterUpload/vsmobilecenterupload.ts]: Done

File: Tasks/VSMobileCenterUpload/utils.ts [hunk 2]
```diff
@@ -98,17 +100,35 @@ export function createZipFile(zipStream: NodeJS.ReadableStream, filename: string
...
+        return matches[0];
+    }
+
+    return null;
+}
+
+export function resolvePaths(pattern: string, continueOnError: boolean = false, packParentFolder: boolean = false): string[] {
     if (pattern) {
-        let matches: string[] = tl.glob(pattern);
+        let matches = tl.glob(pattern);
```
Comment (DavidStaheli) [inline Tasks/VSMobileCenterUpload/utils.ts]: Why remove type info (`: string[]`)?

File: Tasks/VSMobileCenterUpload/utils.ts [hunk 3]
```diff
@@ -119,16 +139,66 @@ export function resolveSinglePath(pattern: string, continueOnError: boolean): st
...
+            if (pathSegmants[i] !== commonParent[i]) {
+                // Segments i, i+1, etc. cannot be a part of the resulting common path.
+                cutTail(commonParent, commonParent.length - i);
+                break;
+            }
+        }
+    })
+
+    return commonParent.join("/");
+}
```
Comment (DavidStaheli) [inline Tasks/VSMobileCenterUpload/utils.ts]: Can we also support Windows-style paths with backslashes ("\")?

File: Tasks/VSMobileCenterUpload/utils.ts [hunk 4]
```diff
@@ -119,16 +139,66 @@ export function resolveSinglePath(pattern: string, continueOnError: boolean): st
...
+        return removeDuplicates(matches.map(v => packParentFolder ? path.dirname(v) : v));
+    }
+
+    return null;
+}
+
+export function removeDuplicates(list: string[]): string[] {
+    let unique: string[] = [];
+
+    list.forEach(s => {
```
Comment (yacaovsnc) [inline Tasks/VSMobileCenterUpload/utils.ts]: Not tested, not sure if works.  Arrays in JavaScript are associative arrays, I think we can avoid the "find" call if you just push each element in list as a key in the array, and get all unique values from `Object.keys(array)` call.  Again I didn't test this yet.

File: Tasks/VSMobileCenterUpload/vsmobilecenterupload.ts [hunk 5]
```diff
@@ -6,7 +6,7 @@ import fs = require('fs');
 
 import { ToolRunner } from 'vsts-task-lib/toolrunner';
 
-var utils = require('./utils.js');
+import utils = require('./utils');
```
Comment (yacaovsnc) [inline Tasks/VSMobileCenterUpload/vsmobilecenterupload.ts]: I had some problem mocking this if it's "import"ed.  You can try and see if mock works.

File: Tasks/VSMobileCenterUpload/vsmobilecenterupload.ts [hunk 6]
```diff
@@ -331,7 +330,32 @@ async function run() {
...
             continueIfSymbolsNotFound = true;
         }
-        let symbolsPath = utils.checkAndFixFilePath(utils.resolveSinglePath(symbolsPathPattern, continueIfSymbolsNotFound), "symbolsPath", continueIfSymbolsNotFound);
+
+        let symbolsPaths: string[] = []
+        if (symbolsType === "dsymPath") {
+            let dsymPaths = utils.resolvePaths(symbolsPathPattern, continueIfSymbolsNotFound, packParentFolder);
+            
+            dsymPaths.forEach(dsymFolder => 
+            {
```
Comment (yacaovsnc) [inline Tasks/VSMobileCenterUpload/vsmobilecenterupload.ts]: The opening brace should be on the same line?

File: Tasks/VSMobileCenterUpload/utils.ts (hunk 4 above)
Comment (yacaovsnc) [inline Tasks/VSMobileCenterUpload/utils.ts]:
```
> a
[ 'a/b', 'b', 'a', 'a/b' ]
> b = []
[]
> a.forEach(function(s) {b[s] = 1})
undefined
> Object.keys(b)
[ 'a/b', 'b', 'a' ]
```
Seems working.

Comment (AlexRukhlin) [inline Tasks/VSMobileCenterUpload/utils.ts]:
```
> var array = []
undefined
> array.push("a/b/c")
1
> array.push("a/b/c/d")
2
> array.push("a/b/c")
3
> Object.keys(array)
[ '0', '1', '2' ]
```

File: Tasks/VSMobileCenterUpload/utils.ts (hunk 3 above)
Comment (AlexRukhlin) [inline Tasks/VSMobileCenterUpload/utils.ts]: It looks like we do not need that. Paths processed in this method always have forward slashes (even in Windows).

File: Tasks/VSMobileCenterUpload/utils.ts (hunk 2 above)
Comment (AlexRukhlin) [inline Tasks/VSMobileCenterUpload/utils.ts]: The type is defined by the right side of assignment.

File: Tasks/VSMobileCenterUpload/vsmobilecenterupload.ts (hunk 6 above)
Comment (AlexRukhlin) [inline Tasks/VSMobileCenterUpload/vsmobilecenterupload.ts]: Done

File: Tasks/VSMobileCenterUpload/utils.ts (hunk 4 above)
Comment (AlexRukhlin) [inline Tasks/VSMobileCenterUpload/utils.ts]: Done

File: Tasks/VSMobileCenterUpload/vsmobilecenterupload.ts (hunk 5 above)
Comment (AlexRukhlin) [inline Tasks/VSMobileCenterUpload/vsmobilecenterupload.ts]: Done

Comment (AlexRukhlin) [inline Tasks/VSMobileCenterUpload/vsmobilecenterupload.ts]: Done

**Pull request merged**